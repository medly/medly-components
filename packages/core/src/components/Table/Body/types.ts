export interface TableBodyProps {
    rowCursor?: number | string;
    onRowNavigated?: (data: { [key: string]: any }) => void;
    setUniqueIds: React.Dispatch<React.SetStateAction<any[]>>;
    selectedRowIds: Array<number | string>;
    onRowSelection: (id: number | string) => void;
    showShadowAfterFrozenElement: boolean;
    setSelectAllDisableState: React.Dispatch<React.SetStateAction<boolean>>;
    onGroupedRowSelection: (id: number[] | string[]) => void;
}
