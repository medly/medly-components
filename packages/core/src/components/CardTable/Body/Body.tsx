import React, { FC, useCallback, useMemo } from 'react';
import Text from '../../Text';
import Cell from './Cell';
import Row from './Row';
import { NoResult } from './Row/Row.styled';
import { Props } from './types';

export const Body: FC<Props> = React.memo(props => {
    const { data, columns, onRowClick, rowClickDisableKey, uniqueKeyName, withWhiteBackground } = props;

    const gridTemplateColumns = useMemo(() => columns.reduce((acc, curr) => acc.concat(`${curr.fraction || 1}fr `), ''), [columns]),
        handleRowClick = useCallback(
            (rowData: any) => {
                return onRowClick && !rowData[rowClickDisableKey!] ? () => onRowClick(rowData) : undefined;
            },
            [onRowClick, rowClickDisableKey]
        );

    if (data.length === 0) {
        return (
            <NoResult withWhiteBackground={withWhiteBackground} gridTemplateColumns={gridTemplateColumns}>
                <Text textVariant="body2" textWeight="Strong">
                    No Result
                </Text>
            </NoResult>
        );
    }

    return (
        <tbody>
            {data.map((row, index) => (
                <Row
                    key={row[uniqueKeyName!] || index}
                    onClick={handleRowClick(row)}
                    withWhiteBackground={withWhiteBackground}
                    isRowClickDisabled={row[rowClickDisableKey!]}
                    gridTemplateColumns={gridTemplateColumns}
                >
                    {columns.map(column => (
                        <Cell
                            key={column.field}
                            isRowClickDisabled={row[rowClickDisableKey!]}
                            data={row[column.field]}
                            column={column}
                            rowId={row[uniqueKeyName!]}
                            rowData={row}
                        />
                    ))}
                </Row>
            ))}
        </tbody>
    );
});
Body.defaultProps = {
    uniqueKeyName: 'id',
    rowClickDisableKey: ''
};
Body.displayName = 'Body';
