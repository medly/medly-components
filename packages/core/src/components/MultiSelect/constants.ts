import { KeyBindings } from './types';

export const defaultKeyBindings: KeyBindings = {
    up: 'ArrowUp',
    down: 'ArrowDown',
    select: ' ',
    close: 'Escape'
};
