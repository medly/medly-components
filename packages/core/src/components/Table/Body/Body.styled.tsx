import styled from 'styled-components';

export const TBody = styled('tbody')`
    display: block;
    min-width: fit-content;
`;
