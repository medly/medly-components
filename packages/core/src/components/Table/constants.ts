import { KeyBindings, TableColumnConfig } from './types';

export const rowActionsColumnConfig: TableColumnConfig = {
    title: '',
    field: 'row-actions',
    frozen: true
};

export const loadingBodyData = [{ id: 'loading1' }, { id: 'loading2' }, { id: 'loading3' }, { id: 'loading4' }, { id: 'loading5' }];

export const defaultKeyBindings: KeyBindings = {
    up: 'ArrowUp',
    down: 'ArrowDown',
    expandRow: 'ArrowRight',
    collapseRow: 'ArrowLeft',
    clickRow: 'Enter',
    selectRow: ' '
};
