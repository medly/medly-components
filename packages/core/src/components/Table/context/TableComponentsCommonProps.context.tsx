import { createContext } from 'react';
import { TableProps } from '../types';

export const TableComponentsCommonPropsContext = createContext<
    TableProps & {
        rowIdentifier: string;
        size: Required<TableProps>['size'];
        isGroupedTable: boolean;
        tableRef: React.RefObject<HTMLTableElement>;
        hiddenDivRef: React.MutableRefObject<any>;
        addColumnMaxSize: (field: string, value: number) => void;
        keyBindings: Required<TableProps>['keyBindings'];
    }
    // @ts-ignore
>({});
