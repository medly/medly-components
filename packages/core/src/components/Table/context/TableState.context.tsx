import React from 'react';
import { TableState } from '../types';

export const TableStateContext = React.createContext<[TableState, React.Dispatch<React.SetStateAction<TableState>>]>([{}, () => null]);
