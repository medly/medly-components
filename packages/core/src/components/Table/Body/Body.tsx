import { useKeyPress } from '@medly-components/utils';
import { FC, memo, useContext, useEffect, useMemo, useState } from 'react';
import { TableComponentsCommonPropsContext } from '../context';
import { getGridTemplateColumns } from '../helpers';
import Minimap from '../Minimap';
import { TBody } from './Body.styled';
import GroupedRow from './GroupedRow';
import Row from './Row';
import { NoResultCell, NoResultRow } from './Row/Row.styled';
import { TableBodyProps } from './types';

const Body: FC<TableBodyProps> = memo(props => {
    const {
            data,
            groupBy,
            rowIdentifier,
            showRowWithCardStyle,
            noResultRow,
            noResultRowText,
            tableRef,
            withMinimap,
            columns,
            size,
            keyBindings
        } = useContext(TableComponentsCommonPropsContext),
        {
            selectedRowIds,
            onRowSelection,
            onGroupedRowSelection,
            setUniqueIds,
            rowCursor: customCursor,
            onRowNavigated,
            ...restProps
        } = props,
        [cursorState, setCursor] = useState(-1),
        /* since minimap is positioned sticky with respect to the tbody, tbody should have full table width otherwise minimap positioning fails */
        tableVisibleWidth = tableRef.current?.clientWidth ?? 0,
        minimapDimensionDeps = useMemo(() => [columns], [columns]),
        isUpKeyPressed = useKeyPress(keyBindings.up!, false, tableRef),
        isDownKeyPressed = useKeyPress(keyBindings.down!, false, tableRef),
        isExpansionKeyPressed = useKeyPress(keyBindings.expandRow!, false, tableRef),
        isCollapseKeyPressed = useKeyPress(keyBindings.collapseRow!, false, tableRef);

    useEffect(() => {
        !customCursor && data.length && isUpKeyPressed && setCursor(prevState => (prevState > 0 ? prevState - 1 : prevState));
    }, [isUpKeyPressed, data, customCursor]);

    useEffect(() => {
        !customCursor &&
            data.length &&
            isDownKeyPressed &&
            setCursor(prevState => (prevState < data.length - 1 ? prevState + 1 : prevState));
    }, [isDownKeyPressed, data, customCursor]);

    const cursor = customCursor ?? cursorState;

    return (
        <TBody>
            {data.length === 0 && (
                <NoResultRow
                    showRowWithCardStyle={showRowWithCardStyle}
                    gridTemplateColumns={getGridTemplateColumns(columns)}
                    withMinimap={withMinimap}
                >
                    <NoResultCell width={tableVisibleWidth} tableSize={size}>
                        {noResultRow ? noResultRow : noResultRowText || 'No result'}
                    </NoResultCell>
                </NoResultRow>
            )}
            {data.map((row, index) => {
                const identifier = (groupBy ? row[groupBy] : row[rowIdentifier]) || index,
                    uniqueId = isNaN(Number(identifier)) ? index : identifier;
                return groupBy ? (
                    <GroupedRow
                        id={uniqueId}
                        key={identifier}
                        titleRowData={row}
                        setUniqueIds={setUniqueIds}
                        selectedTitleRowIds={selectedRowIds}
                        onTitleRowSelection={onRowSelection}
                        {...{ ...restProps, onGroupedRowSelection }}
                    />
                ) : (
                    <Row
                        id={uniqueId}
                        key={identifier}
                        data={row}
                        onRowNavigated={onRowNavigated}
                        selectedRowIds={selectedRowIds}
                        onRowSelection={onRowSelection}
                        isNavigated={index === cursor}
                        isRowExpandedFromKeyboard={isExpansionKeyPressed && index === cursor}
                        isRowCollapsedFromKeyboard={isCollapseKeyPressed && index === cursor}
                        {...restProps}
                    />
                );
            })}
            {withMinimap && <Minimap tableRef={tableRef} minimapDimensionDeps={minimapDimensionDeps} />}
        </TBody>
    );
});
Body.displayName = 'Body';
export default Body;
