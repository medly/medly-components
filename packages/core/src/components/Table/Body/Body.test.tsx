import { render, screen } from '@test-utils';
import { RefObject } from 'react';
import { defaultKeyBindings } from '../constants';
import { TableComponentsCommonPropsContext } from '../context';
import testColumns from '../docs/columns';
import testData from '../docs/data';
import Body from './Body';

const tableRef = {
    current: {
        scrollWidth: 300,
        offsetWidth: 200,
        clientWidth: 200,
        scrollLeft: 1,
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        getBoundingClientRect: jest.fn().mockImplementation(() => ({
            bottom: 100
        })),
        scroll: jest.fn()
    }
} as unknown as RefObject<HTMLTableElement>;

const setUniqueIds = jest.fn(),
    onRowSelection = jest.fn(),
    setSelectAllDisableState = jest.fn(),
    onGroupedRowSelection = jest.fn(),
    defaultProps = {
        setUniqueIds,
        selectedRowIds: [1],
        onRowSelection,
        showShadowAfterFrozenElement: false,
        setSelectAllDisableState,
        onGroupedRowSelection
    },
    contextValue = {
        columns: testColumns,
        size: 'S' as 'XS' | 'S' | 'M' | 'L',
        rowIdentifier: testColumns[0].field,
        data: testData,
        isGroupedTable: false,
        tableRef,
        hiddenDivRef: tableRef,
        addColumnMaxSize: jest.fn(),
        keyBindings: {
            ...defaultKeyBindings
        }
    },
    renderer = (props = defaultProps, context = contextValue) =>
        render(
            <TableComponentsCommonPropsContext.Provider value={context}>
                <Body {...props} />
            </TableComponentsCommonPropsContext.Provider>
        );

describe('Row', () => {
    it('should render tbody properly with default props', () => {
        const { container } = renderer();
        expect(screen.getByText(testData[0].name)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should render tbody properly with `No result` content', () => {
        renderer(defaultProps, { ...contextValue, data: [] });
        expect(screen.getByText('No result')).toBeInTheDocument();
    });
});
