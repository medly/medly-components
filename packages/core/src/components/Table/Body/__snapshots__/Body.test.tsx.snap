// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row should render tbody properly with default props 1`] = `
.c0 {
  display: block;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
}

.c4 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c5 {
  display: grid;
  position: relative;
  opacity: 1;
  width: 100%;
  height: 100%;
  padding: 0;
}

.c2 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 1;
  padding: 0.8rem 1.6rem;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c2 .c3 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c6 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 1;
  padding: 0.8rem 1.6rem;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c6 .c3 {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c1 {
  display: grid;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  outline: none;
  cursor: inherit;
}

.c1.c1:hover {
  z-index: 2;
}

.c1:nth-child(odd),
.c1:nth-child(odd) > * {
  background-color: #F7F8F9;
  color: #13181D;
}

.c1:nth-child(even),
.c1:nth-child(even) > * {
  background-color: #ffffff;
  color: #13181D;
}

.c7 {
  display: grid;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: -webkit-fit-content;
  min-width: -moz-fit-content;
  min-width: fit-content;
  outline: none;
  cursor: inherit;
}

.c7.c7:hover {
  z-index: 2;
}

.c7:nth-child(odd),
.c7:nth-child(odd) > * {
  background-color: #D7E3F6;
  color: #13181D;
}

.c7:nth-child(even),
.c7:nth-child(even) > * {
  background-color: #D7E3F6;
  color: #13181D;
}

<div>
  <tbody
    class="c0"
  >
    <tr
      class="c1"
      style="grid-template-columns: undefined minmax(0px, 0fr) undefined undefined undefined;"
      tabindex="0"
    >
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Albus Percival Wulfric Brian Dumbledore"
        >
          Albus Percival Wulfric Brian Dumbledore
        </span>
      </td>
      <td
        class="c5"
        style="grid-template-columns: undefined undefined;"
      >
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="6"
          >
            6
          </span>
        </div>
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="4"
          >
            4
          </span>
        </div>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="12"
        >
          12
        </span>
      </td>
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Yes"
        >
          Yes
        </span>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="5"
        >
          5
        </span>
      </td>
    </tr>
    <tr
      class="c7"
      style="grid-template-columns: undefined minmax(0px, 0fr) undefined undefined undefined;"
      tabindex="1"
    >
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Mary May"
        >
          Mary May
        </span>
      </td>
      <td
        class="c5"
        style="grid-template-columns: undefined undefined;"
      >
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="4"
          >
            4
          </span>
        </div>
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="7"
          >
            7
          </span>
        </div>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="1"
        >
          1
        </span>
      </td>
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Yes"
        >
          Yes
        </span>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="4"
        >
          4
        </span>
      </td>
    </tr>
    <tr
      class="c1"
      style="grid-template-columns: undefined minmax(0px, 0fr) undefined undefined undefined;"
      tabindex="2"
    >
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Christine Lobowski"
        >
          Christine Lobowski
        </span>
      </td>
      <td
        class="c5"
        style="grid-template-columns: undefined undefined;"
      >
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="4"
          >
            4
          </span>
        </div>
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="7"
          >
            7
          </span>
        </div>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="42"
        >
          42
        </span>
      </td>
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Yes"
        >
          Yes
        </span>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="4"
        >
          4
        </span>
      </td>
    </tr>
    <tr
      class="c1"
      style="grid-template-columns: undefined minmax(0px, 0fr) undefined undefined undefined;"
      tabindex="3"
    >
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Brendon Philips"
        >
          Brendon Philips
        </span>
      </td>
      <td
        class="c5"
        style="grid-template-columns: undefined undefined;"
      >
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="10"
          >
            10
          </span>
        </div>
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="2"
          >
            2
          </span>
        </div>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="125"
        >
          125
        </span>
      </td>
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Yes"
        >
          Yes
        </span>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="4.5"
        >
          4.5
        </span>
      </td>
    </tr>
    <tr
      class="c1"
      style="grid-template-columns: undefined minmax(0px, 0fr) undefined undefined undefined;"
      tabindex="4"
    >
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Margret Marmajuke"
        >
          Margret Marmajuke
        </span>
      </td>
      <td
        class="c5"
        style="grid-template-columns: undefined undefined;"
      >
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="7"
          >
            7
          </span>
        </div>
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="5"
          >
            5
          </span>
        </div>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="16"
        >
          16
        </span>
      </td>
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Yes"
        >
          Yes
        </span>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="4"
        >
          4
        </span>
      </td>
    </tr>
    <tr
      class="c1"
      style="grid-template-columns: undefined minmax(0px, 0fr) undefined undefined undefined;"
      tabindex="5"
    >
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Van Ng"
        >
          Van Ng
        </span>
      </td>
      <td
        class="c5"
        style="grid-template-columns: undefined undefined;"
      >
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="9"
          >
            9
          </span>
        </div>
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="2"
          >
            2
          </span>
        </div>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="37"
        >
          37
        </span>
      </td>
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="No"
        >
          No
        </span>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="4"
        >
          4
        </span>
      </td>
    </tr>
    <tr
      class="c1"
      style="grid-template-columns: undefined minmax(0px, 0fr) undefined undefined undefined;"
      tabindex="6"
    >
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="Duc Ng"
        >
          Duc Ng
        </span>
      </td>
      <td
        class="c5"
        style="grid-template-columns: undefined undefined;"
      >
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="10"
          >
            10
          </span>
        </div>
        <div
          class="c6"
        >
          <span
            class="c3 c4"
            title="3"
          >
            3
          </span>
        </div>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="37"
        >
          37
        </span>
      </td>
      <td
        class="c2"
      >
        <span
          class="c3 c4"
          title="No"
        >
          No
        </span>
      </td>
      <td
        class="c6"
      >
        <span
          class="c3 c4"
          title="4"
        >
          4
        </span>
      </td>
    </tr>
  </tbody>
</div>
`;
